### ISSUE TITLE
Create access throughput service 

### ISSUE THREAD

**Issue by autumnfound** [EclipseFdn/marketplace-rest-api#9]
A new service will be created that maintains a list of external connections created to the API, using a piece of identifiable information and a timestamp to store a record of access. This service will be bound to CDI to allow for access throughout the application. As this service will be multithreaded, access will need to synchronized on mutable elements.

On request of the throughput service, any expired entries should be removed before any operations (such as adding new entries or checking for abuse) are performed.

If a user triggers the an unsuccessful request too many times after limit has been reached, then  a temporary blacklist can be triggered to reduce likelihood of abuse of the API. There should be the ability to weight different types of requests to trigger timeouts faster for different types of requests.

The following properties will be created within the microprofile for the application:  

- eclipse.throughput.max: The maximum number of accesses allowed per period
- eclipse.throughput.period: The period of time to record access for.
- eclipse.throughput.burst.max: The number of requests that should be allowed within a burst period.
- eclipse.throughput.burst.period: The period of time that defines a "burst" of requests 
- eclipse.throughput.burst.weight: Allows for users abusing the API within a short period of time to trigger a timeout sooner by increasing the weight of unsuccessful burst requests.
- eclipse.throughput.whitelist: Whitelist of personally identifiable tokens that should always be allowed access
- eclipse.throughput.blacklist: Blacklist of personally identifiable tokens that should always be restricted access
- eclipse.throughput.timeout.max: The maximum number of requests to allow to the API
- eclipse.throughput.timeout.period: The timeout period for too many requests.

**Comment by autumnfound** [comment]
Additionally, there should be a method that returns the time until the current user can make requests if they are currently limited or blacklisted.

Secondly, when a successful request occurs for a user, any state that tracks abuse should be cleared.

**Comment by autumnfound** [comment]
For implementation, there are a few options available to discuss:

## Redis cached rate limit implementation
https://github.com/mokies/ratelimitj/tree/master/ratelimitj-redis

**Pros:**

- Apache license
- Cluster/cloud ready w/ Redis caching

**Cons:**

- Requires Redis to be configured
- Non-mature offering 

## RateLimitter Class (Guava)
https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/RateLimiter.java

**Pros:**

- Apache license
- Part of mature + trusted library

**Cons:**

- Non-distributed implementation

## Bucket4J
https://github.com/vladimir-bukhtoyarov/bucket4j

**Pros:**

- Distributed implementation support
- Apache license

**Cons:**

- Would require separate GRID cache to be configured (limited integration w/ Redis ([102](https://github.com/vladimir-bukhtoyarov/bucket4j/pull/102)))

I'm leaning towards an implementation of ratelimitj-redis and adding redis to the stack as it would be the best for performance and cloud-native implementation goals. Any feedback is appreciated and welcome!

**Comment by mbarbero** [comment]
Rate limiting should not be implemented at the application/service level, otherwise you will just end up re-implement the same stuff over and over again for each app/service. Rate limiting is better handled at a "gateway"/"proxy" level. 

The most simple way to do it is to just use nginx. There are more elaborated solution in the form of API gateways, but let's wait to have more services to orchestrate before diving in this rabbit hole.

**Comment by autumnfound** [comment]
The goal of rebuilding this API was to improve on the previous iteration and to add modern features that weren't possible or feasible in the previous version. Part of this solution is providing information to the user regarding rate limiting. This is usually provided to the user in the form of response headers much like those that [GitHub provide](https://developer.github.com/v3/rate_limit/).

Looking at the solutions you mentioned, it looks like we can achieve this with an API gateway, but not in NGINX (cursory search, so I could be missing an esoteric plugin or something haha!).

**Comment by mbarbero** [comment]
Fair enough about the limitation of nginx if we really want to provide such headers. IMO, it's still a good/quick start.

In all cases, I urge you to not implement it by yourself inside the service/app :)
